modules/renter/downloadcache.go

```diff
@@ -14,21 +14,35 @@ import (
 // TODO this won't be necessary anymore once we have partial downloads.
 func (udc *unfinishedDownloadChunk) addChunkToCache(data []byte) {
 	if udc.download.staticDestinationType != destinationTypeSeekStream {
-		// We only cache streaming chunks since browsers and media players tend to only request a few kib at once when streaming data. That way we can prevent scheduling the same chunk for download over and over.
+		// We only cache streaming chunks since browsers and media players tend
+		// to only request a few kib at once when streaming data. That way we can
+		// prevent scheduling the same chunk for download over and over.
 		return
 	}
 	udc.cacheMu.Lock()
+	defer udc.cacheMu.Unlock()
+
 	// Prune cache if necessary.
-	// TODO insteado of deleting a 'random' key, delete the
-	// least-recently-accessed element of the cache.
-	for key := range udc.chunkCache {
-		if len(udc.chunkCache) < downloadCacheSize {
-			break
+	for len(udc.chunkCache) >= downloadCacheSize {
+		var oldestKey string
+		oldestTime := time.Now()
+
+		// TODO: turn this from a structure where you loop over every element
+		// (O(n) per access) to a min heap (O(log n) per access).
+		// currently not a issue due to cache size remaining small (<20)
+		for id, chunk := range udc.chunkCache {
+			if chunk.lastAccess.Before(oldestTime) {
+				oldestTime = chunk.lastAccess
+				oldestKey = id
+			}
 		}
-		delete(udc.chunkCache, key)
+		delete(udc.chunkCache, oldestKey)
+	}
```

Comment:
instead of a min heap, we could use a bit of extra memory to sort and then delete. The runtime is O(n log n), which is slightly worse than the minheap's O(m log n), but not terribly so. Something like:
```go
if len(udc.chunkCache) > downloadCacheSize {
	cacheIDs := make([]string, 0, len(udc.chunkCache))
	for id := range udc.chunkCache {
		cacheIDs = append(cacheIDs, id)
	}
	sort.Slice(cacheIDs, func(i, j int) bool {
		ci := udc.chunkCache[cacheIDs[i]]
		cj := udc.chunkCache[cacheIDs[j]]
		return ci.lastAccess.Before(cj.lastAccess)
	})
	for len(udc.chunkCache) > downloadCacheSize {
		delete(udc.chunkCache, cacheIDs[0])
		cacheIDs = cacheIDs[1:]
	}
}
```

although, come to think of it, don't we do this pruning every time we insert a chunk? That is, won't the `for len(udc.chunkCache) >= downloadCacheSize` loop only ever need to run once? If so, a min heap makes sense (because it will persist across calls), but sorting doesn't.

Reply:
The for loop doesn't really hurt but yeah you are right. It should only ever be executed once which makes the heap O(log n) for pruning vs O(n log n).